evap/staff/views.py

```diff
@@ -677,48 +678,91 @@ def course_participant_import(request, semester_id, course_id):
-            file_content = get_import_file_content_or_raise(request.user.id)
-            imported_users, success_messages, warnings, __ = UserImporter.process(file_content, test_run)
-            delete_import_file(request.user.id)
+    participant_test_passed = import_file_exists(request.user.id, 'participant')
+    contributor_test_passed = import_file_exists(request.user.id, 'contributor')
+    # casting warnings to a normal dict is necessary for the template to iterate over it.
+    return render(request, "staff_course_person_import.html", dict(semester=semester, course=course,
+        participant_excel_form=participant_excel_form, participant_copy_form=participant_copy_form,
+        contributor_excel_form=contributor_excel_form, contributor_copy_form=contributor_copy_form,
+        success_messages=success_messages, warnings=dict(warnings), errors=errors,
+        participant_test_passed=participant_test_passed, contributor_test_passed=contributor_test_passed))
+
+
+def helper_course_person_import(request, excel_form, copy_form, course, semester_id):
+    operation = request.POST.get('operation')
+
+    redirect_result = None
+    success_messages = []
+    warnings = {}
+    errors = []
+    magic = 'participant' if 'participants' in operation else 'contributor'
+
+    if 'test' in operation:
+        delete_import_file(request.user.id, magic)  # remove old files if still exist
+        excel_form.excel_file_required = True
+        if excel_form.is_valid():
+            excel_file = excel_form.cleaned_data['excel_file']
+            file_content = excel_file.read()
+            # if on a test run, the process method will not return a list of users that would be
+            # imported so there is currently no way to show how many users would be imported.
+            __, success_messages, warnings, errors = UserImporter.process(file_content, True)
+            if not errors:
+                save_import_file(excel_file, request.user.id, magic)
+
+    elif 'import' in operation:
+        file_content = get_import_file_content_or_raise(request.user.id, magic)
+        imported_users, success_messages, warnings, __ = UserImporter.process(file_content, False)
+        delete_import_file(request.user.id, magic)
+        if magic == 'participant':
             course.participants.add(*imported_users)
             success_messages.append(("{} Participants added to course {}").format(len(imported_users), course.name))
-            forward_messages(request, success_messages, warnings)
-            return redirect('staff:semester_view', semester_id)
-
-        elif copy_run:
-            copy_form.course_selection_required = True
-            if copy_form.is_valid():
-                import_course = copy_form.cleaned_data['course']
+        else:
+            helper_add_contributors(imported_users, course)
+            success_messages.append(("{} Contributors added to course {}").format(len(imported_users), course.name))
```

Comment:
could you check how that would look like if you moved these 6 lines into a helper method, re-used that method in the `copy` branch of this view (you'd still need to determine `imported_users` outside the helper)?